src/mlir/dialect/Typechecker.cc

```diff
@@ -0,0 +1,162 @@
+    // not trivial. We achieve it by matching against the type of the operator()
+    // method. The matching part works by specializing a rule_traits struct for
+    // member methods. The trick was borrowed from MLIR's walk methods in
+    // llvm/include/mlir/IR/Visitors.h
+    //
+    // If the lambda is too generic, eg. `[](auto x, auto y) { ... }` taking the
+    // address of the operator() method will fail and so will compilation. This
+    // is intentional as we wouldn't know in that case what the target of the
+    // dyncast should be.
+    template<typename F>
+    struct rule_traits;
+    template<typename F>
+    Rule(F f)
+      ->Rule<
+        typename rule_traits<decltype(&F::operator())>::Left,
+        typename rule_traits<decltype(&F::operator())>::Right,
+        F>;
+
+    template<typename C, typename A1, typename A2>
+    struct rule_traits<bool (C::*)(A1, A2) const>
+    {
+      using Left = A1;
+      using Right = A2;
+    };
+
+    template<typename C, typename A1, typename A2>
+    struct rule_traits<bool (C::*)(A1, A2)>
+    {
+      using Left = A1;
+      using Right = A2;
+    };
+
+    /// From a set of callable objects, create a single closure which calls each
+    /// of them in succession, until one of them works.
+    ///
+    /// The callables `Fs` should have precise (ie. non-auto) argument types,
+    /// such that the deduction guide described earlier works.
+    template<typename... Fs>
+    static constexpr auto combineRules(Fs... rules)
+    {
+      return [=](Type lhs, Type rhs) { return (Rule(rules)(lhs, rhs) || ...); };
+    }
+  }
+
+  /// Set of rules used for subtype-checking. Each rule is checked (in the
+  /// order they are laid out) until one of them works.
+  ///
+  /// `RULES` is a callable object, which takes two Type values and returns a
+  /// boolean.
+  static constexpr auto RULES = detail::combineRules(
+    [](Type left, Type right) { return left == right; },
+    [](JoinType left, Type right) {
+      return llvm::all_of(left.getElements(), [&](Type element) {
+        return isSubtype(element, right);
+      });
+    },
+    [](Type left, MeetType right) {
+      return llvm::all_of(right.getElements(), [&](Type element) {
+        return isSubtype(left, element);
+      });
+    },
+    [](MeetType left, Type right) {
+      return llvm::any_of(left.getElements(), [&](Type element) {
+        return isSubtype(element, right);
+      });
+    },
+    [](Type left, JoinType right) {
+      return llvm::any_of(right.getElements(), [&](Type element) {
+        return isSubtype(left, element);
+      });
+    });
+
+  bool isSubtype(Type lhs, Type rhs)
+  {
+    // TODO: Applying the syntax-directed rules enumerated above is not enough
+    // to recognize certain relationships, such as distributivity:
+    //
+    // (A & (B | C)) <: (A & B) | (A & C)
+    //
+    // For this to work, we would need to apply normalization to the types
```

Comment:
We definitely want to normalise types, or chasing patterns can get quite complicated.

Reply:
Yes, I've got a follow up PR to do just that. I just thought it would be better to break things up.